src/lib/pubkey/frodokem/frodokem.cpp

```diff
@@ -0,0 +1,398 @@
+      }
+
+      const FrodoKEMConstants& constants() const { return m_constants; }
+
+      const FrodoSeedA& seed_a() const { return m_seed_a; }
+
+      const FrodoMatrix& b() const { return m_b; }
+
+      const FrodoPublicKeyHash& hash() const { return m_hash; }
+
+      std::vector<uint8_t> serialize() const {
+         return concat_as<std::vector<uint8_t>>(seed_a(), b().pack(m_constants));
+      }
+
+   private:
+      FrodoKEMConstants m_constants;
+      FrodoSeedA m_seed_a;
+      FrodoMatrix m_b;
+      FrodoPublicKeyHash m_hash;
+};
+
+class FrodoKEM_PrivateKeyInternal {
+   public:
+      FrodoKEM_PrivateKeyInternal(FrodoSeedS s, FrodoMatrix s_trans) :
+            m_s(std::move(s)), m_s_trans(std::move(s_trans)) {}
+
+      const FrodoSeedS& s() const { return m_s; }
+
+      const FrodoMatrix& s_trans() const { return m_s_trans; }
+
+   private:
+      FrodoSeedS m_s;
+      FrodoMatrix m_s_trans;
+};
+
+//
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+//
+
+class Frodo_KEM_Encryptor final : public PK_Ops::KEM_Encryption_with_KDF {
+   public:
+      Frodo_KEM_Encryptor(std::shared_ptr<FrodoKEM_PublicKeyInternal> key, std::string_view kdf) :
+            KEM_Encryption_with_KDF(kdf), m_public_key(std::move(key)) {}
+
+      size_t raw_kem_shared_key_length() const override { return m_public_key->constants().len_sec_bytes(); }
+
+      size_t encapsulated_key_length() const override { return m_public_key->constants().len_ct_bytes(); }
+
+      void raw_kem_encrypt(std::span<uint8_t> out_encapsulated_key,
+                           std::span<uint8_t> out_shared_key,
+                           RandomNumberGenerator& rng) override {
+         const auto& consts = m_public_key->constants();
+         auto& shake = consts.SHAKE_XOF();
+
+         BufferStuffer out_ct_bs(out_encapsulated_key);
+
+         auto c_1 = out_ct_bs.next<FrodoPackedMatrix>(consts.len_packed_b_bytes());
+         auto c_2 = out_ct_bs.next<FrodoPackedMatrix>(consts.len_packed_c_bytes());
+         auto salt = out_ct_bs.next<FrodoSalt>(consts.len_salt_bytes());
+
+         BOTAN_ASSERT_NOMSG(out_ct_bs.full());
+
+         const auto u = rng.random_vec<FrodoPlaintext>(consts.len_sec_bytes());
+         rng.randomize(salt);
+
+         shake.update(m_public_key->hash());
+         shake.update(u);
+         shake.update(salt);
+         const auto seed_se = shake.output<FrodoSeedSE>(consts.len_se_bytes());
+         const auto k = shake.output<FrodoIntermediateSharedSecret>(consts.len_sec_bytes());
+         shake.clear();
+
+         shake.update(consts.encapsulation_domain_separator());
+         shake.update(seed_se);
+
+         const auto s_p =
+            FrodoMatrix::sample(consts,
+                                std::tuple(consts.n_bar(), consts.n()),
+                                shake.output<FrodoSampleR>(sizeof(uint16_t) * consts.n_bar() * consts.n()));
+
+         const auto e_p =
```

Comment:
We have here three repetitions of

```
FrodoMatrix::sample(consts,
    std::tuple(consts.n_bar(), consts.XXX()),
    shake.output<FrodoSampleR>(sizeof(uint16_t) * consts.n_bar() * consts.XXX())
```

I'm not sure if this is the best approach but: at the minimum it seems like a lambda that captures `shake` as a mutable reference and takes `XXX` (either `n_bar` or `n`) as an argument would clean this up.

Reply:
There are three similar invocations during decrypt - maybe a `sample` helper on `FrodoMatrix` or a function in an anon namespace in this file would be better than a lambda which would still have to be duplicated.

Reply:
Added a helper function to `FrodoMatrix`. This nicely reduces complexity, thanks for the tip!